kotest-property/kotest-property-core/src/commonMain/kotlin/io/kotest/property/core/constraints.kt

```diff
@@ -0,0 +1,36 @@
+package io.kotest.property.core
+
+import kotlin.time.Duration
+import kotlin.time.TimeSource
+
+/**
+ * Controls iterations of a property test.
+ */
+fun interface Constraints {
+
+   fun evaluate(context: Iteration): Boolean
+
+   companion object {
+
+      /**
+       * Returns a [Constraints] that executes the property test for a fixed number of iterations.
+       */
+      fun iterations(k: Int) = Constraints { it.iteration < k }
+
+      /**
+       * Returns a [Constraints] that executes the property test for a certain duration.
+       */
+      fun duration(duration: Duration) = object : Constraints {
+         val mark = TimeSource.Monotonic.markNow() + duration // Property tests cannot use virtual time
+         override fun evaluate(context: Iteration): Boolean {
+            return mark.hasNotPassedNow()
+         }
```

Comment:
I can't help but notice that we are beginning to duplcate extension functions against collections and sequences, such as `take`, `filter` etc. 

Maybe we should reimagine the whole concept, so that arbs are sequences, and then we can utilize all the existing extension functions such as `filter` and `take`?

Reply:
I'm not sure we could make it work via sequences. There's more going on, ie with shrinks, that a simple sequence could provide.

Reply:
`Sequence` is just an interface and wrapping an iterator; so we could probably support all the functionality by simply implementing that interface. Even `Iterator` is just an interface, so technically, an Arb could also be an iterator instead. Or both - but that feels wrong.

Reply:
Upon further thought, I'm wondering how much benefit it would add - you can only really benefit from that if you explicitly use generators as sequences then. However, the beauty (especially of the new DSL) is to have _one_ element at a time, and not having to bother with generation process.

However, supporting sequence interfaces (or iterator) might help with special cases or with the beauty of Kotest's internal codebase.